demos/http/http_demo_s3_upload/http_demo_s3_upload.c

```diff
@@ -0,0 +1,561 @@
+ * response headers and body.
+ *
+ * @note This demo shows how the same buffer can be re-used for storing the HTTP
+ * response after the HTTP request is sent out. However, the user can also
+ * decide to use separate buffers for storing the HTTP request and response.
+ */
+static uint8_t userBuffer[ USER_BUFFER_LENGTH ];
+
+/**
+ * @brief Represents header data that will be sent in an HTTP request.
+ */
+static HTTPRequestHeaders_t requestHeaders;
+
+/**
+ * @brief Configurations of the initial request headers that are passed to
+ * #HTTPClient_InitializeRequestHeaders.
+ */
+static HTTPRequestInfo_t requestInfo;
+
+/**
+ * @brief Represents a response returned from an HTTP server.
+ */
+static HTTPResponse_t response;
+
+/**
+ * @brief The host address string extracted from the pre-signed URL.
+ *
+ * @note S3_PRESIGNED_PUT_URL_LENGTH is set as the array length here as the
+ * length of the host name string cannot exceed this value.
+ */
+static char serverHost[ S3_PRESIGNED_PUT_URL_LENGTH ];
+
+/**
+ * @brief The length of the host address found in the pre-signed URL.
+ */
+static size_t serverHostLength;
+
+/**
+ * @brief The location of the path within the pre-signed URL.
+ */
+static const char * pPath;
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Connect to HTTP server with reconnection retries.
+ *
+ * @param[out] pNetworkContext The output parameter to return the created
+ * network context.
+ *
+ * @return EXIT_FAILURE on failure; EXIT_SUCCESS on successful connection.
+ */
+static int32_t connectToServer( NetworkContext_t * pNetworkContext );
+
+/**
+ * @brief Retrieve and verify the size of the S3 object that is specified in
+ * pPath.
+ *
+ * @param[in] pTransportInterface The transport interface for making network
+ * calls.
+ * @param[in] pPath The Request-URI to the objects of interest. This string
+ * should be null-terminated.
+ *
+ * @return The status of the file size acquisition and verification using a GET
+ * request to the server: true on success, false on failure.
+ */
+static bool verifyS3ObjectFileSize( const TransportInterface_t * pTransportInterface,
+                                    const char * pPath );
+
+/**
+ * @brief Send an HTTP PUT request based on a specified path to upload a file,
+ * then print the response received from the server.
+ *
+ * @param[in] pTransportInterface The transport interface for making network
+ * calls.
+ * @param[in] pPath The Request-URI to the objects of interest. This string
+ * should be null-terminated.
+ *
+ * @return The status of the file upload using a PUT request to the server: true
+ * on success, false on failure.
+ */
+static bool uploadS3ObjectFile( const TransportInterface_t * pTransportInterface,
+                                const char * pPath );
+
+/*-----------------------------------------------------------*/
+
+static int32_t connectToServer( NetworkContext_t * pNetworkContext )
```

Comment:
You can separate this out with these parameters.
```suggestion
static int32_t connectToServer( NetworkContext_t * pNetworkContext,
                                const char* rootCaCertPath,
                                const char* alpnProtocol,
                                size_t alpnProtocolLen,
                                uint16_t port,
                                const char* serverHost,
                                size_t serverHostLength )
```

You can parse for the server host address using getUrlAddress() in the demo code, like you did for the download demo.